### PR TITLE
Update best practice guide

### PR DESCRIPTION
**Pull request by mike-solomon**
* Adds an introduction and some context on what you should know to get the most use out of this document.
* Adds specific examples in many sections. I found it difficult to understand a lot of what was being said without these examples so I'm hoping this will be more welcoming for others.
* Adds more direct guidance of actions people should take.
* Breaks up some longer paragraphs into smaller, more readable chunks.
* Expands the state section.
* Removes the simple, JRE-provided types section as we don't support complex types.
* Adds a new date best practice based on a suggestion from Olga.

### CODE REVIEW

File: authoring-recipes/recipe-conventions-and-best-practices.md [hunk 1]
```diff
@@ -4,82 +4,153 @@ description: Conventions to follow, pitfalls to avoid
...
+* Understand how [recipes](/concepts-and-explanations/recipes.md) work at a high level (by using [visitors](/concepts-and-explanations/visitors.md) to operate on diverse and unexpected [LSTs](/concepts-and-explanations/lossless-semantic-trees.md))
+
+## Best Practices
 
 ### Do No Harm
 
-If your Recipe cannot determine that a change is safe to make, such as when a type you're looking for is unavailable, it is always preferable to make no change than to make the wrong change. Relatedly, when making changes always strive to make the most minimal, least invasive version of that change. Keeping changes minimal makes for easy-to-review diffs and higher rates of pull request acceptance. A change that unnecessarily clobbers formatting or is otherwise overly broad will burden code reviewers and drastically reduce the rate at which they accept changes.
+One of the most important conventions to keep in mind when creating recipes is the concept of "Do No Harm". If your recipe cannot determine that a change is safe to me, **it should make no changes rather than making a potentially wrong change**.
```
Comment (tkvangorder) [inline authoring-recipes/recipe-conventions-and-best-practices.md:21]: You can remove "to me"

Comment (tkvangorder) [reply authoring-recipes/recipe-conventions-and-best-practices.md:21]: ...cannot determine a change is safe, **it should.....

File: authoring-recipes/recipe-conventions-and-best-practices.md [hunk 2]
```diff
@@ -4,82 +4,153 @@ description: Conventions to follow, pitfalls to avoid
...
+
+By reducing a recipe's scope and minimizing the number of changes a recipe makes, you'll make it easier for others to review the changes. This, in turn, will result in more pull requests being accepted and more trust in your recipes.
+
+On the other hand, a change that unnecessarily clobbers formatting or is otherwise overly broad in scope will burden code reviewers and drastically reduce the rate at which they accept changes.
 
 {% hint style="success" %}
-RewriteTest helps you to verify that your recipe does not make unnecessary changes by running your recipe in a loop. If you see your change being made many times it is likely your visitor fails to avoid making unnecessary changes.
+[RewriteTest](/authoring-recipes/recipe-testing.md#rewritetest-interface) helps you to verify that your recipe does not make unnecessary changes by running your recipe in a loop. If you see your change being made many times it is likely your visitor fails to avoid making unnecessary changes. See the [cycle section](#stay-single-cycle) for more information.
```
Comment (tkvangorder) [inline authoring-recipes/recipe-conventions-and-best-practices.md:32]: I would replace "running your recipe in a loop" to "running your recipe over multiple cycles"

File: authoring-recipes/recipe-conventions-and-best-practices.md [hunk 3]
```diff
@@ -4,82 +4,153 @@ description: Conventions to follow, pitfalls to avoid
...
+
+Because of that, whenever shared state is needed within a single visitor or recipe, the cursor stack should be used instead of adding state to the execution context.
+
+#### Override the visit method if you need pass state between visitors
+
+If you need to pass state between different visitors in the same recipe, you should override the `visit` function and create a variable whose scope is accessible to your visitors. For example, in the [AddDependency recipe](https://github.com/openrewrite/rewrite/blob/v7.34.3/rewrite-maven/src/main/java/org/openrewrite/maven/AddDependency.java), we create a [scopeByProject Map](https://github.com/openrewrite/rewrite/blob/v7.34.3/rewrite-maven/src/main/java/org/openrewrite/maven/AddDependency.java#L155) that is then shared between a [UsesTypes visitor](https://github.com/openrewrite/rewrite/blob/v7.34.3/rewrite-maven/src/main/java/org/openrewrite/maven/AddDependency.java#L159-L164) and a [MavenVisitor](https://github.com/openrewrite/rewrite/blob/v7.34.3/rewrite-maven/src/main/java/org/openrewrite/maven/AddDependency.java#L191).
```
Comment (tkvangorder) [inline authoring-recipes/recipe-conventions-and-best-practices.md:154]: I would be more explicit here as more than one class has visit methods.  Let's use `Recipe.visit(before, ExecutionContext)`

**Review by tkvangorder — CHANGES_REQUESTED** [review]

**Comment by mike-solomon** [comment]
Thanks for the suggestions @tkvangorder -- fixed all of them.

**Review by tkvangorder — APPROVED** [review]

**Pull request merged**